### DISCUSSION TITLE
Can't deserialize

### DISCUSSION THREAD

**Discussion by Ingeniouz** [peter-kish/gloot#62] [Q&A] [answered]
Hello, I tried to recreate what was written on the readme page about serialization and deserialization, but it doesn't seem to work. Maybe I am doing something wrong? Here is how I try to use it. The file contains this:

> {"node_name":"InventoryGridPlayer","item_protoset":"res://ItemProtoset.tres","items":[{"node_name":"2x2_bomb","protoset":"res://ItemProtoset.tres","prototype_id":"2x2_bomb"}],"size":"(10, 5)"}

The serialization code:

>      if Input.is_action_just_pressed("save_and_close"):
> 		# Serialize the inventory into a JSON string
> 		var inventory: Inventory = get_node("Inventory/InventoryGridPlayer")
> 		var dict: Dictionary = inventory.serialize()
> 		var json = JSON.print(dict)
> 		# Save to file
> 		var file = File.new()
> 		file.open("res://inventory_save.res", File.WRITE)
> 		file.store_string(json)
> 		file.close()
> 

then trying to deserialize the file on launch:

> func _ready():
> 	var file = File.new()
> 	file.open("res://inventory_save.res", File.READ)
> 	var read_file = file.get_as_text()
> 	file.close()
> 	# Deserialize the inventory from a file
> 	var inventory: Inventory = get_node("Inventory/InventoryGridPlayer")
> 	var res: JSONParseResult = JSON.parse(read_file)
> 	if res.error == OK:
> 		var dict = res.result
> 		inventory.deserialize(dict)
> 

or I tried to deserialize on a button press, doesn't work either :(
The target inventory stays empty even if the file contained the json.
Also the Output is giving me this:

> Key 'size' has wrong type! Expected '5', got '4'!

Maybe this is the reason, idk what this is, I was copying everything from the instruction.
Maybe I should refresh the inventory or something?

**Comment by peter-kish** [comment]
Yup, I got the same error. Seems like the issue is in the way Godot serializes and desterilizes `Vector2` fields (and maybe some others too) in a dictionary.

I created and [issue](https://github.com/peter-kish/gloot/issues/63) for this and I'll try to fix it in the following days.

**Reply by Ingeniouz** [reply]
Thanks! I like your plugin very much and hope to use it in my projects.

**Reply by Ingeniouz** [reply]
Btw I don't know how you are planning to solve it, I just had an idea that maybe size_x and size_y instead of Vector2(size_x, size_y) could be the easy way

**Reply by peter-kish** [reply]
That's one way to do it. Another would be to just parse the string that represents the Vector2.

The issue doesn't really seem to be in `InventoryGrid.serialize()`, since it returns a dictionary that has a valid Vector2. The problem is in storing the dictionary as JSON.
JSON doesn't have a Vector2 type built in, so it stores it as a string. When Godot parses that same JSON it has no way of knowing if that string was originally a Vector2 or any other type, so it just keeps it as a string.

**Reply by Ingeniouz** [reply]
Ahh, seems logical. I just tried to dict.size = get_node("Inventory/InventoryGridPlayer").size right before deserializing, and it is working when the item is at the (0,0), but if I try to load another position of an item, it gives an error "Invalid type in function 'add_item_at' in base 'Node (inventory_grid.gd)'. Cannot convert argument 2 from String to Vector2." So it seems there is also an issue with grid_position(x,y). Thank you!

**Reply by peter-kish** [reply]
Yes, there are also other places where Vector2s are serialized. I'll probably need to make sure that the resulting dictionary doesn't contain any types that are not supported by JSON. And that probably means writing custom functions to serialize/deserialize Vector2.

**Reply by Ingeniouz** [reply]
Hey, I googled a bit and maybe I found the way! There was some discussions about storing Vector2, and I tried the solution to use file.store_var(dict) and var read_file = file.get_var() instead of store_string(json) and get_as_text() them. And it seem to work now!

> if Input.is_action_just_pressed("save_and_close"):
		# Serialize the inventory into a JSON string
		var inventory: Inventory = get_node("Inventory/InventoryGridPlayer")
		var dict: Dictionary = inventory.serialize()
		# Save to file
		var file = File.new()
		file.open("res://inventory_save.res", File.WRITE)
		file.store_var(dict)
		file.close()
	if Input.is_action_just_pressed("ui_select"):
		# Deserialize the inventory from a file
		var file = File.new()
		file.open("res://inventory_save.res", File.READ)
		var read_file = file.get_var()
		file.close()
		# Deserialize the inventory from a JSON string
		var inventory: Inventory = get_node("Inventory/InventoryGridPlayer")
		inventory.deserialize(read_file)

**Reply by Ingeniouz** [reply]
I may have skipped the error checking and something else, and the file seems human-unreadable now, but it saves and loads now.

**Reply by peter-kish** [reply]
Looking good, but I'd still like to keep the option to easily serialize/deserialize from JSON.
I'm currently experimenting with storing types that are not supported in JSON as strings using `var2str` and `str2var`.

**Comment by peter-kish** [comment, marked as answer]
Should be fixed in 115eaa77fd0436c628825f4f36083856238be704

**Reply by Ingeniouz** [reply]
Thank you!
